File: backend/app/api/v1/routes/health.py
"""Liveness, readiness, and Prometheus metrics endpoints."""

from __future__ import annotations

from fastapi import APIRouter, Depends, Response, status
from prometheus_client import CONTENT_TYPE_LATEST, generate_latest
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db_session
from app.dependencies.services import get_vector_store
from app.integrations.qdrant import QdrantVectorStore

router = APIRouter(tags=["Operations"])


@router.get("/health/live")
async def liveness() -> dict[str, str]:
    """Return process liveness."""
    return {"status": "ok"}


@router.get("/health/ready")
async def readiness(
    response: Response,
    session: AsyncSession = Depends(get_db_session),
    vector_store: QdrantVectorStore = Depends(get_vector_store),
) -> dict[str, object]:
    """Check PostgreSQL and Qdrant readiness."""
    checks: dict[str, str] = {}
    try:
        await session.execute(text("SELECT 1"))
        checks["database"] = "ok"
    except Exception:
        checks["database"] = "failed"
    try:
        await vector_store.client.get_collections()
        checks["qdrant"] = "ok"
    except Exception:
        checks["qdrant"] = "failed"
    ready = all(value == "ok" for value in checks.values())
    if not ready:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
    return {"status": "ok" if ready else "degraded", "checks": checks}


@router.get("/metrics", include_in_schema=False)
async def metrics() -> Response:
    """Expose Prometheus metrics."""
    return Response(content=generate_latest(), media_type=CONTENT_TYPE_LATEST)

File: backend/src/routes/health.js
import { Router } from "express";
import { register } from "prom-client";
import { pool } from "../db.js";
import { vectorStore } from "../integrations/qdrant.js";

const router = Router();

router.get("/health/live", (_req, res) => {
  res.json({ status: "ok" });
});

router.get("/health/ready", async (_req, res) => {
  const checks = {};
  try {
    await pool.query("SELECT 1");
    checks.database = "ok";
  } catch {
    checks.database = "failed";
  }
  try {
    await vectorStore.ready();
    checks.qdrant = "ok";
  } catch {
    checks.qdrant = "failed";
  }
  const ready = Object.values(checks).every((value) => value === "ok");
  res.status(ready ? 200 : 503).json({
    status: ready ? "ok" : "degraded",
    checks,
  });
});

router.get("/metrics", async (_req, res) => {
  res.set("Content-Type", register.contentType);
  res.send(await register.metrics());
});

export default router;
